02_bubble/alfredxiao/main.go

```diff
@@ -0,0 +1,51 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func makeCopy(ns []int) []int {
+	return append([]int{}, ns...)
+}
+
+func bubble(s []int) []int {
+	copy := makeCopy(s)
+	n := len(copy)
+	for i := n - 1; i > 0; i-- {
+		for j := 0; j < i; j++ {
+			maybeSwap(copy, j, j+1)
+		}
+	}
+	return copy
```

Comment:
Nit: Personally I'd inline `len` and `maybeSwap` - only change it if you see value/ agree. Additionaly you can add an early bail here too:

```go
	for i := len(s); i > 0; i-- {
		swapped := false
		for j := 1; j < i; j++ {
			if s[j-1] > s[j] {
				s[j-1], s[j] = s[j], s[j-1]
				swapped = true
			}
		}
		if !swapped {
			// If no elements were swapped, the slice is sorted so we can stop here
			break
		}
	}